src/core/dataset.ts

```diff
@@ -4,9 +4,20 @@ module Plottable {
 
 export type DatasetCallback = (dataset: Dataset) => void;
 
+export class KeyFunctions {
+  protected static counter: number = 0;
+  public static noConstancy: (d: any, i: number) => any = (d: any, i: number) => {
+    return KeyFunctions.counter++;
+  };
```

Comment:
If you call ```selection.bind(data)``` without supplying a key function, d3 defaults to using the index of the datum in the data array as the key - ie same as the ```useIndex``` function here.  This is why, currently, if you simply replace the data with itself - ```dataset.data(data)``` - no animation would occur without the ```animateOnNextRender``` logic. That forces everything back to a base position so it can then move back to where it was before.